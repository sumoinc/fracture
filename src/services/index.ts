export * from "./data-service";
export * from "./operation";
export * from "./resource-attribute";
export * from "./resource";
export * from "./service";
export * from "./structure-attribute";
export * from "./structure";
