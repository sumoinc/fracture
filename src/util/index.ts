export * from "./determineBuildId";
export * from "./test-util";
