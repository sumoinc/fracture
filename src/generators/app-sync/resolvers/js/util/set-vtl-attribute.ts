import { ResourceAttributeGenerator } from "../../../../../services/resource-attribute";
import { StructureAttribute } from "../../../../../services/structure-attribute";
import { TypescriptStrategy } from "../../../../ts/strategy";

export const setVtlAttribute = (attribute: StructureAttribute) => {
  const strategy = TypescriptStrategy.of(attribute.project);
  const generator = attribute.generator;
  const name = strategy.formatAttributeName(attribute.name);
  const shortName = strategy.formatAttributeName(attribute.shortName);

  // non-generated
  if (generator === ResourceAttributeGenerator.NONE) {
    return `"${shortName}": $util.dynamodb.toDynamoDBJson($ctx.args.input.${name}),`;
    // generated values
  } else {
    switch (generator) {
      case ResourceAttributeGenerator.CURRENT_DATE_TIME_STAMP:
        return `"${shortName}": "$util.time.nowISO8601()"`;
      case ResourceAttributeGenerator.GUID:
        return `"${shortName}": "$util.autoId()"`;
      case ResourceAttributeGenerator.TYPE:
        return `"${shortName}": "unknown"`;

      default:
        throw new Error(`Unknown generator: ${generator}`);
    }
  }
};
