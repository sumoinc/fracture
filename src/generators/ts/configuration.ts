import { Service } from "../../services/service";
import { GeneratedTypescriptFile } from "../generated-typescript-file";

export class Configuration extends GeneratedTypescriptFile {
  constructor(public readonly project: Service) {
    super(project, "configuration.ts");
  }

  preSynthesize(): void {
    super.preSynthesize();

    this.addLine(
      `export const configurations = ${JSON.stringify(
        this.project.config(),
        null,
        2
      )}`
    );
  }
}
