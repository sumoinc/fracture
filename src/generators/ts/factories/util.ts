import { ValueOf } from "type-fest";
import { ResourceAttributeType } from "../../../services";

export const JsType = {
  STRING: "string",
  NUMBER: "number",
  BOOLEAN: "boolean",
  ARRAY: "array",
  MAP: "map",
  OTHER: "other",
} as const;

export const jsType = (
  resourceAttributeType: ValueOf<typeof ResourceAttributeType> | string
) => {
  switch (resourceAttributeType) {
    case ResourceAttributeType.GUID:
    case ResourceAttributeType.STRING:
    case ResourceAttributeType.EMAIL:
    case ResourceAttributeType.PHONE:
    case ResourceAttributeType.URL:
    case ResourceAttributeType.DATE:
    case ResourceAttributeType.TIME:
    case ResourceAttributeType.DATE_TIME:
    case ResourceAttributeType.JSON:
    case ResourceAttributeType.IPADDRESS:
      return JsType.STRING;
    case ResourceAttributeType.INT:
    case ResourceAttributeType.FLOAT:
    case ResourceAttributeType.TIMESTAMP:
    case ResourceAttributeType.COUNT:
    case ResourceAttributeType.AVERAGE:
    case ResourceAttributeType.SUM:
      return JsType.NUMBER;
    case ResourceAttributeType.BOOLEAN:
      return JsType.BOOLEAN;
    case ResourceAttributeType.ARRAY:
      return JsType.ARRAY;
    case ResourceAttributeType.MAP:
      return JsType.MAP;
    default:
      return JsType.OTHER;
  }
};
