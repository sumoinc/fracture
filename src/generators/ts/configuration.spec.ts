import { Configuration } from "./configuration";
import { Resource } from "../../services";
import { synthFile, testDataService } from "../../util/test-util";

test("Smoke test", () => {
  const service = testDataService();
  new Resource(service, {
    name: "user",
    attributeOptions: [
      {
        name: "name",
      },
    ],
  });

  new Resource(service, {
    name: "widget",
    tenantEnabled: true,
    attributeOptions: [
      {
        name: "name",
      },
    ],
  });
  new Configuration(service);

  const content = synthFile(service, "src/configuration.ts");
  expect(content).toBeTruthy();
  expect(content).toMatchSnapshot();
  // console.log(content);
});
