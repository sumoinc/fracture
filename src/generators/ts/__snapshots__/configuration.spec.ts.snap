// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Smoke test 1`] = `
"// ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".

export const configurations = {
  \\"name\\": \\"my-service\\",
  \\"tenantEnabled\\": false,
  \\"dynamo\\": {
    \\"name\\": \\"my-service-table\\",
    \\"pk\\": {
      \\"name\\": \\"pk\\",
      \\"attributeType\\": \\"S\\",
      \\"keyType\\": \\"HASH\\"
    },
    \\"sk\\": {
      \\"name\\": \\"sk\\",
      \\"attributeType\\": \\"S\\",
      \\"keyType\\": \\"RANGE\\"
    },
    \\"idx\\": {
      \\"name\\": \\"idx\\",
      \\"attributeType\\": \\"S\\",
      \\"keyType\\": \\"NONE\\"
    },
    \\"keyGsi\\": {
      \\"name\\": \\"key\\",
      \\"pk\\": {
        \\"name\\": \\"pk\\",
        \\"attributeType\\": \\"S\\",
        \\"keyType\\": \\"HASH\\"
      },
      \\"sk\\": {
        \\"name\\": \\"sk\\",
        \\"attributeType\\": \\"S\\",
        \\"keyType\\": \\"RANGE\\"
      },
      \\"type\\": \\"PRIMARY\\"
    },
    \\"lookupGsi\\": {
      \\"name\\": \\"loookup\\",
      \\"pk\\": {
        \\"name\\": \\"sk\\",
        \\"attributeType\\": \\"S\\",
        \\"keyType\\": \\"RANGE\\"
      },
      \\"sk\\": {
        \\"name\\": \\"idx\\",
        \\"attributeType\\": \\"S\\",
        \\"keyType\\": \\"NONE\\"
      },
      \\"type\\": \\"SECONDARY\\"
    },
    \\"gsi\\": [
      {
        \\"name\\": \\"key\\",
        \\"pk\\": {
          \\"name\\": \\"pk\\",
          \\"attributeType\\": \\"S\\",
          \\"keyType\\": \\"HASH\\"
        },
        \\"sk\\": {
          \\"name\\": \\"sk\\",
          \\"attributeType\\": \\"S\\",
          \\"keyType\\": \\"RANGE\\"
        },
        \\"type\\": \\"PRIMARY\\"
      },
      {
        \\"name\\": \\"loookup\\",
        \\"pk\\": {
          \\"name\\": \\"sk\\",
          \\"attributeType\\": \\"S\\",
          \\"keyType\\": \\"RANGE\\"
        },
        \\"sk\\": {
          \\"name\\": \\"idx\\",
          \\"attributeType\\": \\"S\\",
          \\"keyType\\": \\"NONE\\"
        },
        \\"type\\": \\"SECONDARY\\"
      }
    ]
  },
  \\"resources\\": [
    {
      \\"name\\": \\"user\\",
      \\"shortName\\": \\"user\\",
      \\"pluralName\\": \\"users\\",
      \\"comments\\": [],
      \\"tenantEnabled\\": false,
      \\"dynamoTable\\": {
        \\"name\\": \\"my-service-table\\",
        \\"pk\\": {
          \\"name\\": \\"pk\\",
          \\"attributeType\\": \\"S\\",
          \\"keyType\\": \\"HASH\\"
        },
        \\"sk\\": {
          \\"name\\": \\"sk\\",
          \\"attributeType\\": \\"S\\",
          \\"keyType\\": \\"RANGE\\"
        },
        \\"idx\\": {
          \\"name\\": \\"idx\\",
          \\"attributeType\\": \\"S\\",
          \\"keyType\\": \\"NONE\\"
        },
        \\"keyGsi\\": {
          \\"name\\": \\"key\\",
          \\"pk\\": {
            \\"name\\": \\"pk\\",
            \\"attributeType\\": \\"S\\",
            \\"keyType\\": \\"HASH\\"
          },
          \\"sk\\": {
            \\"name\\": \\"sk\\",
            \\"attributeType\\": \\"S\\",
            \\"keyType\\": \\"RANGE\\"
          },
          \\"type\\": \\"PRIMARY\\"
        },
        \\"lookupGsi\\": {
          \\"name\\": \\"loookup\\",
          \\"pk\\": {
            \\"name\\": \\"sk\\",
            \\"attributeType\\": \\"S\\",
            \\"keyType\\": \\"RANGE\\"
          },
          \\"sk\\": {
            \\"name\\": \\"idx\\",
            \\"attributeType\\": \\"S\\",
            \\"keyType\\": \\"NONE\\"
          },
          \\"type\\": \\"SECONDARY\\"
        },
        \\"gsi\\": [
          {
            \\"name\\": \\"key\\",
            \\"pk\\": {
              \\"name\\": \\"pk\\",
              \\"attributeType\\": \\"S\\",
              \\"keyType\\": \\"HASH\\"
            },
            \\"sk\\": {
              \\"name\\": \\"sk\\",
              \\"attributeType\\": \\"S\\",
              \\"keyType\\": \\"RANGE\\"
            },
            \\"type\\": \\"PRIMARY\\"
          },
          {
            \\"name\\": \\"loookup\\",
            \\"pk\\": {
              \\"name\\": \\"sk\\",
              \\"attributeType\\": \\"S\\",
              \\"keyType\\": \\"RANGE\\"
            },
            \\"sk\\": {
              \\"name\\": \\"idx\\",
              \\"attributeType\\": \\"S\\",
              \\"keyType\\": \\"NONE\\"
            },
            \\"type\\": \\"SECONDARY\\"
          }
        ]
      },
      \\"attributes\\": [
        {
          \\"name\\": \\"pk\\",
          \\"shortName\\": \\"pk\\",
          \\"required\\": true,
          \\"comments\\": [
            \\"Partition Key for this record.\\"
          ],
          \\"type\\": \\"String\\",
          \\"management\\": \\"SYSTEM_MANAGED\\",
          \\"visibility\\": \\"HIDDEN\\",
          \\"identifier\\": \\"NONE\\",
          \\"compositionSources\\": [
            \\"id\\"
          ],
          \\"compositionsSeperator\\": \\"#\\",
          \\"createGenerator\\": \\"Composition\\",
          \\"readGenerator\\": \\"Composition\\",
          \\"updateGenerator\\": \\"Composition\\",
          \\"deleteGenerator\\": \\"Composition\\"
        },
        {
          \\"name\\": \\"sk\\",
          \\"shortName\\": \\"sk\\",
          \\"required\\": true,
          \\"comments\\": [
            \\"Sort Key for this record.\\"
          ],
          \\"type\\": \\"String\\",
          \\"management\\": \\"SYSTEM_MANAGED\\",
          \\"visibility\\": \\"HIDDEN\\",
          \\"identifier\\": \\"NONE\\",
          \\"compositionSources\\": [
            \\"type\\",
            \\"version\\"
          ],
          \\"compositionsSeperator\\": \\"#\\",
          \\"createGenerator\\": \\"Composition\\",
          \\"readGenerator\\": \\"Composition\\",
          \\"updateGenerator\\": \\"Composition\\",
          \\"deleteGenerator\\": \\"Composition\\"
        },
        {
          \\"name\\": \\"lookup\\",
          \\"shortName\\": \\"idx\\",
          \\"required\\": true,
          \\"comments\\": [
            \\"Lookup value for this record.\\"
          ],
          \\"type\\": \\"String\\",
          \\"management\\": \\"SYSTEM_MANAGED\\",
          \\"visibility\\": \\"HIDDEN\\",
          \\"identifier\\": \\"NONE\\",
          \\"compositionSources\\": [],
          \\"compositionsSeperator\\": \\" \\",
          \\"createGenerator\\": \\"Composition\\",
          \\"readGenerator\\": \\"Composition\\",
          \\"updateGenerator\\": \\"Composition\\",
          \\"deleteGenerator\\": \\"Composition\\"
        },
        {
          \\"name\\": \\"id\\",
          \\"shortName\\": \\"id\\",
          \\"required\\": true,
          \\"comments\\": [
            \\"Identifier for this record.\\"
          ],
          \\"type\\": \\"String\\",
          \\"management\\": \\"SYSTEM_MANAGED\\",
          \\"visibility\\": \\"USER_VISIBLE\\",
          \\"identifier\\": \\"PRIMARY\\",
          \\"compositionSources\\": [],
          \\"compositionsSeperator\\": \\"#\\",
          \\"createGenerator\\": \\"Guid\\",
          \\"readGenerator\\": \\"None\\",
          \\"updateGenerator\\": \\"None\\",
          \\"deleteGenerator\\": \\"None\\"
        },
        {
          \\"name\\": \\"type\\",
          \\"shortName\\": \\"t\\",
          \\"required\\": true,
          \\"comments\\": [
            \\"Type of record.\\"
          ],
          \\"type\\": \\"String\\",
          \\"management\\": \\"SYSTEM_MANAGED\\",
          \\"visibility\\": \\"USER_VISIBLE\\",
          \\"identifier\\": \\"NONE\\",
          \\"compositionSources\\": [],
          \\"compositionsSeperator\\": \\"#\\",
          \\"createGenerator\\": \\"Type\\",
          \\"readGenerator\\": \\"None\\",
          \\"updateGenerator\\": \\"None\\",
          \\"deleteGenerator\\": \\"None\\"
        },
        {
          \\"name\\": \\"version\\",
          \\"shortName\\": \\"v\\",
          \\"required\\": true,
          \\"comments\\": [
            \\"Version for record.\\"
          ],
          \\"type\\": \\"String\\",
          \\"management\\": \\"SYSTEM_MANAGED\\",
          \\"visibility\\": \\"USER_VISIBLE\\",
          \\"identifier\\": \\"NONE\\",
          \\"compositionSources\\": [],
          \\"compositionsSeperator\\": \\"#\\",
          \\"createGenerator\\": \\"VersionDateTimeStamp\\",
          \\"readGenerator\\": \\"None\\",
          \\"updateGenerator\\": \\"None\\",
          \\"deleteGenerator\\": \\"None\\"
        },
        {
          \\"name\\": \\"created-timestamp\\",
          \\"shortName\\": \\"ct\\",
          \\"required\\": true,
          \\"comments\\": [
            \\"The timestamp representing when this record was created.\\"
          ],
          \\"type\\": \\"String\\",
          \\"management\\": \\"SYSTEM_MANAGED\\",
          \\"visibility\\": \\"USER_VISIBLE\\",
          \\"identifier\\": \\"NONE\\",
          \\"compositionSources\\": [],
          \\"compositionsSeperator\\": \\"#\\",
          \\"createGenerator\\": \\"CurrentDateTimeStamp\\",
          \\"readGenerator\\": \\"None\\",
          \\"updateGenerator\\": \\"CurrentDateTimeStamp\\",
          \\"deleteGenerator\\": \\"None\\"
        },
        {
          \\"name\\": \\"modified-timestamp\\",
          \\"shortName\\": \\"mt\\",
          \\"required\\": true,
          \\"comments\\": [
            \\"The timestamp representing when this record was last modified.\\"
          ],
          \\"type\\": \\"String\\",
          \\"management\\": \\"SYSTEM_MANAGED\\",
          \\"visibility\\": \\"USER_VISIBLE\\",
          \\"identifier\\": \\"NONE\\",
          \\"compositionSources\\": [],
          \\"compositionsSeperator\\": \\"#\\",
          \\"createGenerator\\": \\"None\\",
          \\"readGenerator\\": \\"None\\",
          \\"updateGenerator\\": \\"CurrentDateTimeStamp\\",
          \\"deleteGenerator\\": \\"None\\"
        },
        {
          \\"name\\": \\"deleted-timestamp\\",
          \\"shortName\\": \\"dt\\",
          \\"required\\": true,
          \\"comments\\": [
            \\"The timestamp representing when this record was marked as deleted.\\"
          ],
          \\"type\\": \\"String\\",
          \\"management\\": \\"SYSTEM_MANAGED\\",
          \\"visibility\\": \\"HIDDEN\\",
          \\"identifier\\": \\"NONE\\",
          \\"compositionSources\\": [],
          \\"compositionsSeperator\\": \\"#\\",
          \\"createGenerator\\": \\"None\\",
          \\"readGenerator\\": \\"None\\",
          \\"updateGenerator\\": \\"None\\",
          \\"deleteGenerator\\": \\"CurrentDateTimeStamp\\"
        },
        {
          \\"name\\": \\"name\\",
          \\"shortName\\": \\"name\\",
          \\"required\\": true,
          \\"comments\\": [],
          \\"type\\": \\"String\\",
          \\"management\\": \\"USER_MANAGED\\",
          \\"visibility\\": \\"USER_VISIBLE\\",
          \\"identifier\\": \\"NONE\\",
          \\"compositionSources\\": [],
          \\"compositionsSeperator\\": \\"#\\",
          \\"createGenerator\\": \\"None\\",
          \\"readGenerator\\": \\"None\\",
          \\"updateGenerator\\": \\"None\\",
          \\"deleteGenerator\\": \\"None\\"
        }
      ],
      \\"dataStructure\\": {
        \\"name\\": \\"user\\",
        \\"comments\\": [
          \\"Public facing data structure for this record.\\"
        ],
        \\"attributes\\": [
          {
            \\"name\\": \\"id\\",
            \\"shortName\\": \\"id\\",
            \\"type\\": \\"String\\",
            \\"management\\": \\"SYSTEM_MANAGED\\",
            \\"visibility\\": \\"USER_VISIBLE\\",
            \\"comments\\": [
              \\"Identifier for this record.\\"
            ],
            \\"required\\": true,
            \\"generator\\": \\"None\\"
          },
          {
            \\"name\\": \\"type\\",
            \\"shortName\\": \\"t\\",
            \\"type\\": \\"String\\",
            \\"management\\": \\"SYSTEM_MANAGED\\",
            \\"visibility\\": \\"USER_VISIBLE\\",
            \\"comments\\": [
              \\"Type of record.\\"
            ],
            \\"required\\": true,
            \\"generator\\": \\"None\\"
          },
          {
            \\"name\\": \\"version\\",
            \\"shortName\\": \\"v\\",
            \\"type\\": \\"String\\",
            \\"management\\": \\"SYSTEM_MANAGED\\",
            \\"visibility\\": \\"USER_VISIBLE\\",
            \\"comments\\": [
              \\"Version for record.\\"
            ],
            \\"required\\": true,
            \\"generator\\": \\"None\\"
          },
          {
            \\"name\\": \\"created-timestamp\\",
            \\"shortName\\": \\"ct\\",
            \\"type\\": \\"String\\",
            \\"management\\": \\"SYSTEM_MANAGED\\",
            \\"visibility\\": \\"USER_VISIBLE\\",
            \\"comments\\": [
              \\"The timestamp representing when this record was created.\\"
            ],
            \\"required\\": true,
            \\"generator\\": \\"None\\"
          },
          {
            \\"name\\": \\"modified-timestamp\\",
            \\"shortName\\": \\"mt\\",
            \\"type\\": \\"String\\",
            \\"management\\": \\"SYSTEM_MANAGED\\",
            \\"visibility\\": \\"USER_VISIBLE\\",
            \\"comments\\": [
              \\"The timestamp representing when this record was last modified.\\"
            ],
            \\"required\\": true,
            \\"generator\\": \\"None\\"
          },
          {
            \\"name\\": \\"name\\",
            \\"shortName\\": \\"name\\",
            \\"type\\": \\"String\\",
            \\"management\\": \\"USER_MANAGED\\",
            \\"visibility\\": \\"USER_VISIBLE\\",
            \\"comments\\": [],
            \\"required\\": true,
            \\"generator\\": \\"None\\"
          }
        ]
      }
    },
    {
      \\"name\\": \\"widget\\",
      \\"shortName\\": \\"widget\\",
      \\"pluralName\\": \\"widgets\\",
      \\"comments\\": [],
      \\"tenantEnabled\\": true,
      \\"dynamoTable\\": {
        \\"name\\": \\"my-service-table\\",
        \\"pk\\": {
          \\"name\\": \\"pk\\",
          \\"attributeType\\": \\"S\\",
          \\"keyType\\": \\"HASH\\"
        },
        \\"sk\\": {
          \\"name\\": \\"sk\\",
          \\"attributeType\\": \\"S\\",
          \\"keyType\\": \\"RANGE\\"
        },
        \\"idx\\": {
          \\"name\\": \\"idx\\",
          \\"attributeType\\": \\"S\\",
          \\"keyType\\": \\"NONE\\"
        },
        \\"keyGsi\\": {
          \\"name\\": \\"key\\",
          \\"pk\\": {
            \\"name\\": \\"pk\\",
            \\"attributeType\\": \\"S\\",
            \\"keyType\\": \\"HASH\\"
          },
          \\"sk\\": {
            \\"name\\": \\"sk\\",
            \\"attributeType\\": \\"S\\",
            \\"keyType\\": \\"RANGE\\"
          },
          \\"type\\": \\"PRIMARY\\"
        },
        \\"lookupGsi\\": {
          \\"name\\": \\"loookup\\",
          \\"pk\\": {
            \\"name\\": \\"sk\\",
            \\"attributeType\\": \\"S\\",
            \\"keyType\\": \\"RANGE\\"
          },
          \\"sk\\": {
            \\"name\\": \\"idx\\",
            \\"attributeType\\": \\"S\\",
            \\"keyType\\": \\"NONE\\"
          },
          \\"type\\": \\"SECONDARY\\"
        },
        \\"gsi\\": [
          {
            \\"name\\": \\"key\\",
            \\"pk\\": {
              \\"name\\": \\"pk\\",
              \\"attributeType\\": \\"S\\",
              \\"keyType\\": \\"HASH\\"
            },
            \\"sk\\": {
              \\"name\\": \\"sk\\",
              \\"attributeType\\": \\"S\\",
              \\"keyType\\": \\"RANGE\\"
            },
            \\"type\\": \\"PRIMARY\\"
          },
          {
            \\"name\\": \\"loookup\\",
            \\"pk\\": {
              \\"name\\": \\"sk\\",
              \\"attributeType\\": \\"S\\",
              \\"keyType\\": \\"RANGE\\"
            },
            \\"sk\\": {
              \\"name\\": \\"idx\\",
              \\"attributeType\\": \\"S\\",
              \\"keyType\\": \\"NONE\\"
            },
            \\"type\\": \\"SECONDARY\\"
          }
        ]
      },
      \\"attributes\\": [
        {
          \\"name\\": \\"pk\\",
          \\"shortName\\": \\"pk\\",
          \\"required\\": true,
          \\"comments\\": [
            \\"Partition Key for this record.\\"
          ],
          \\"type\\": \\"String\\",
          \\"management\\": \\"SYSTEM_MANAGED\\",
          \\"visibility\\": \\"HIDDEN\\",
          \\"identifier\\": \\"NONE\\",
          \\"compositionSources\\": [
            \\"tenant-id\\",
            \\"id\\"
          ],
          \\"compositionsSeperator\\": \\"#\\",
          \\"createGenerator\\": \\"Composition\\",
          \\"readGenerator\\": \\"Composition\\",
          \\"updateGenerator\\": \\"Composition\\",
          \\"deleteGenerator\\": \\"Composition\\"
        },
        {
          \\"name\\": \\"sk\\",
          \\"shortName\\": \\"sk\\",
          \\"required\\": true,
          \\"comments\\": [
            \\"Sort Key for this record.\\"
          ],
          \\"type\\": \\"String\\",
          \\"management\\": \\"SYSTEM_MANAGED\\",
          \\"visibility\\": \\"HIDDEN\\",
          \\"identifier\\": \\"NONE\\",
          \\"compositionSources\\": [
            \\"tenant-id\\",
            \\"type\\",
            \\"version\\"
          ],
          \\"compositionsSeperator\\": \\"#\\",
          \\"createGenerator\\": \\"Composition\\",
          \\"readGenerator\\": \\"Composition\\",
          \\"updateGenerator\\": \\"Composition\\",
          \\"deleteGenerator\\": \\"Composition\\"
        },
        {
          \\"name\\": \\"lookup\\",
          \\"shortName\\": \\"idx\\",
          \\"required\\": true,
          \\"comments\\": [
            \\"Lookup value for this record.\\"
          ],
          \\"type\\": \\"String\\",
          \\"management\\": \\"SYSTEM_MANAGED\\",
          \\"visibility\\": \\"HIDDEN\\",
          \\"identifier\\": \\"NONE\\",
          \\"compositionSources\\": [],
          \\"compositionsSeperator\\": \\" \\",
          \\"createGenerator\\": \\"Composition\\",
          \\"readGenerator\\": \\"Composition\\",
          \\"updateGenerator\\": \\"Composition\\",
          \\"deleteGenerator\\": \\"Composition\\"
        },
        {
          \\"name\\": \\"tenant-id\\",
          \\"shortName\\": \\"tid\\",
          \\"required\\": true,
          \\"comments\\": [
            \\"Tenant identifier for this record.\\"
          ],
          \\"type\\": \\"String\\",
          \\"management\\": \\"SYSTEM_MANAGED\\",
          \\"visibility\\": \\"HIDDEN\\",
          \\"identifier\\": \\"NONE\\",
          \\"compositionSources\\": [],
          \\"compositionsSeperator\\": \\"#\\",
          \\"createGenerator\\": \\"Tenant\\",
          \\"readGenerator\\": \\"Tenant\\",
          \\"updateGenerator\\": \\"Tenant\\",
          \\"deleteGenerator\\": \\"Tenant\\"
        },
        {
          \\"name\\": \\"id\\",
          \\"shortName\\": \\"id\\",
          \\"required\\": true,
          \\"comments\\": [
            \\"Identifier for this record.\\"
          ],
          \\"type\\": \\"String\\",
          \\"management\\": \\"SYSTEM_MANAGED\\",
          \\"visibility\\": \\"USER_VISIBLE\\",
          \\"identifier\\": \\"PRIMARY\\",
          \\"compositionSources\\": [],
          \\"compositionsSeperator\\": \\"#\\",
          \\"createGenerator\\": \\"Guid\\",
          \\"readGenerator\\": \\"None\\",
          \\"updateGenerator\\": \\"None\\",
          \\"deleteGenerator\\": \\"None\\"
        },
        {
          \\"name\\": \\"type\\",
          \\"shortName\\": \\"t\\",
          \\"required\\": true,
          \\"comments\\": [
            \\"Type of record.\\"
          ],
          \\"type\\": \\"String\\",
          \\"management\\": \\"SYSTEM_MANAGED\\",
          \\"visibility\\": \\"USER_VISIBLE\\",
          \\"identifier\\": \\"NONE\\",
          \\"compositionSources\\": [],
          \\"compositionsSeperator\\": \\"#\\",
          \\"createGenerator\\": \\"Type\\",
          \\"readGenerator\\": \\"None\\",
          \\"updateGenerator\\": \\"None\\",
          \\"deleteGenerator\\": \\"None\\"
        },
        {
          \\"name\\": \\"version\\",
          \\"shortName\\": \\"v\\",
          \\"required\\": true,
          \\"comments\\": [
            \\"Version for record.\\"
          ],
          \\"type\\": \\"String\\",
          \\"management\\": \\"SYSTEM_MANAGED\\",
          \\"visibility\\": \\"USER_VISIBLE\\",
          \\"identifier\\": \\"NONE\\",
          \\"compositionSources\\": [],
          \\"compositionsSeperator\\": \\"#\\",
          \\"createGenerator\\": \\"VersionDateTimeStamp\\",
          \\"readGenerator\\": \\"None\\",
          \\"updateGenerator\\": \\"None\\",
          \\"deleteGenerator\\": \\"None\\"
        },
        {
          \\"name\\": \\"created-timestamp\\",
          \\"shortName\\": \\"ct\\",
          \\"required\\": true,
          \\"comments\\": [
            \\"The timestamp representing when this record was created.\\"
          ],
          \\"type\\": \\"String\\",
          \\"management\\": \\"SYSTEM_MANAGED\\",
          \\"visibility\\": \\"USER_VISIBLE\\",
          \\"identifier\\": \\"NONE\\",
          \\"compositionSources\\": [],
          \\"compositionsSeperator\\": \\"#\\",
          \\"createGenerator\\": \\"CurrentDateTimeStamp\\",
          \\"readGenerator\\": \\"None\\",
          \\"updateGenerator\\": \\"CurrentDateTimeStamp\\",
          \\"deleteGenerator\\": \\"None\\"
        },
        {
          \\"name\\": \\"modified-timestamp\\",
          \\"shortName\\": \\"mt\\",
          \\"required\\": true,
          \\"comments\\": [
            \\"The timestamp representing when this record was last modified.\\"
          ],
          \\"type\\": \\"String\\",
          \\"management\\": \\"SYSTEM_MANAGED\\",
          \\"visibility\\": \\"USER_VISIBLE\\",
          \\"identifier\\": \\"NONE\\",
          \\"compositionSources\\": [],
          \\"compositionsSeperator\\": \\"#\\",
          \\"createGenerator\\": \\"None\\",
          \\"readGenerator\\": \\"None\\",
          \\"updateGenerator\\": \\"CurrentDateTimeStamp\\",
          \\"deleteGenerator\\": \\"None\\"
        },
        {
          \\"name\\": \\"deleted-timestamp\\",
          \\"shortName\\": \\"dt\\",
          \\"required\\": true,
          \\"comments\\": [
            \\"The timestamp representing when this record was marked as deleted.\\"
          ],
          \\"type\\": \\"String\\",
          \\"management\\": \\"SYSTEM_MANAGED\\",
          \\"visibility\\": \\"HIDDEN\\",
          \\"identifier\\": \\"NONE\\",
          \\"compositionSources\\": [],
          \\"compositionsSeperator\\": \\"#\\",
          \\"createGenerator\\": \\"None\\",
          \\"readGenerator\\": \\"None\\",
          \\"updateGenerator\\": \\"None\\",
          \\"deleteGenerator\\": \\"CurrentDateTimeStamp\\"
        },
        {
          \\"name\\": \\"name\\",
          \\"shortName\\": \\"name\\",
          \\"required\\": true,
          \\"comments\\": [],
          \\"type\\": \\"String\\",
          \\"management\\": \\"USER_MANAGED\\",
          \\"visibility\\": \\"USER_VISIBLE\\",
          \\"identifier\\": \\"NONE\\",
          \\"compositionSources\\": [],
          \\"compositionsSeperator\\": \\"#\\",
          \\"createGenerator\\": \\"None\\",
          \\"readGenerator\\": \\"None\\",
          \\"updateGenerator\\": \\"None\\",
          \\"deleteGenerator\\": \\"None\\"
        }
      ],
      \\"dataStructure\\": {
        \\"name\\": \\"widget\\",
        \\"comments\\": [
          \\"Public facing data structure for this record.\\"
        ],
        \\"attributes\\": [
          {
            \\"name\\": \\"id\\",
            \\"shortName\\": \\"id\\",
            \\"type\\": \\"String\\",
            \\"management\\": \\"SYSTEM_MANAGED\\",
            \\"visibility\\": \\"USER_VISIBLE\\",
            \\"comments\\": [
              \\"Identifier for this record.\\"
            ],
            \\"required\\": true,
            \\"generator\\": \\"None\\"
          },
          {
            \\"name\\": \\"type\\",
            \\"shortName\\": \\"t\\",
            \\"type\\": \\"String\\",
            \\"management\\": \\"SYSTEM_MANAGED\\",
            \\"visibility\\": \\"USER_VISIBLE\\",
            \\"comments\\": [
              \\"Type of record.\\"
            ],
            \\"required\\": true,
            \\"generator\\": \\"None\\"
          },
          {
            \\"name\\": \\"version\\",
            \\"shortName\\": \\"v\\",
            \\"type\\": \\"String\\",
            \\"management\\": \\"SYSTEM_MANAGED\\",
            \\"visibility\\": \\"USER_VISIBLE\\",
            \\"comments\\": [
              \\"Version for record.\\"
            ],
            \\"required\\": true,
            \\"generator\\": \\"None\\"
          },
          {
            \\"name\\": \\"created-timestamp\\",
            \\"shortName\\": \\"ct\\",
            \\"type\\": \\"String\\",
            \\"management\\": \\"SYSTEM_MANAGED\\",
            \\"visibility\\": \\"USER_VISIBLE\\",
            \\"comments\\": [
              \\"The timestamp representing when this record was created.\\"
            ],
            \\"required\\": true,
            \\"generator\\": \\"None\\"
          },
          {
            \\"name\\": \\"modified-timestamp\\",
            \\"shortName\\": \\"mt\\",
            \\"type\\": \\"String\\",
            \\"management\\": \\"SYSTEM_MANAGED\\",
            \\"visibility\\": \\"USER_VISIBLE\\",
            \\"comments\\": [
              \\"The timestamp representing when this record was last modified.\\"
            ],
            \\"required\\": true,
            \\"generator\\": \\"None\\"
          },
          {
            \\"name\\": \\"name\\",
            \\"shortName\\": \\"name\\",
            \\"type\\": \\"String\\",
            \\"management\\": \\"USER_MANAGED\\",
            \\"visibility\\": \\"USER_VISIBLE\\",
            \\"comments\\": [],
            \\"required\\": true,
            \\"generator\\": \\"None\\"
          }
        ]
      }
    }
  ],
  \\"operations\\": []
}"
`;
